org/mozilla/jss/tests/JSSProvider.java

```diff
@@ -0,0 +1,85 @@
+package org.mozilla.jss.tests;
+
+import java.util.*;
+import java.security.*;
+import javax.crypto.*;
+import javax.net.ssl.*;
+
+import org.mozilla.jss.*;
+import org.mozilla.jss.pkcs11.*;
+
+public class JSSProvider {
+    public static String BASE_PACKAGE = "org.mozilla.jss.provider.";
+
+    public static void shouldProvide(Provider p, String algo, String _type) {
+        String item = (String) p.get(algo);
+        assert(item != null);
+        assert(item.startsWith(BASE_PACKAGE));
+        assert(item.endsWith("." + _type));
+    }
+
+    public static void shouldNotProvide(Provider p, String algo) {
+        assert(p.get(algo) == null);
+    }
+
+    public static void main(String[] args) throws Exception {
+        Provider p = Security.getProvider("Mozilla-JSS");
+        System.out.println(p);
+
+        CryptoManager.initialize(args[0]);
+        CryptoManager cm = CryptoManager.getInstance();
+        cm.setPasswordCallback(new FilePasswordCallback(args[1]));
+
+        p = Security.getProvider("Mozilla-JSS");
+        System.out.println(p);
+
+        // Validate assumptions about how provider.get() works:
+        //
+        // Something which by no means should ever be implemented returns
+        // null.
+        shouldNotProvide(p, "Charlie's Chocolate Factory");
+
+        // Validate that our provider implements certain interfaces we care
+        // about. In particular, ensure they're overriden as part of our
+        // org.mozilla.jss.provider package, and that the class name matches
+        // what we expect our implementation to be.
+        shouldProvide(p, "KeyManagerFactory.NssX509", "JSSKeyManagerFactory");
+
+        // Validate that our provider is "default"; that is, when we get an
+        // algorithm instance without explicitly providing the provider, we
+        // end up with an instance from our provider.
+        Mac m = Mac.getInstance("HmacSHA512");
+        assert(m.getProvider().getName() == p.getName());
```

Comment:
This only works if they are pointing to the same String object, right? Is that the assumption that we want to test here, or should we do something like:
```
assert(m.getProvider().getName().equals(p.getName()));
```

Reply:
D'uh, sorry. This still gets me.